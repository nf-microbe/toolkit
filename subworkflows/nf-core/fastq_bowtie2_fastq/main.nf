// Import modules
include { BOWTIE2_BUILD } from '../../../modules/nf-core/bowtie2/build/main'
include { BOWTIE2_ALIGN } from '../../../modules/nf-core/bowtie2/align/main'

// Run subworkflow
workflow FASTQ_BOWTIE2_FASTQ {

    take:
    fastq_gz        // channel: [ [ meta ], [ reads_1.fastq.gz, reads_2.fastq.gz ] ] (MANDATORY)
    fasta_gz        // val: /path/to/fasta.fasta.gz (OPTIONAL)
    bt2_index       // val: /path/to/host_index.bt2 (OPTIONAL)

    main:
    ch_versions = Channel.empty()

    // prepare fasta and bowtie2 index
    if (fasta_gz && bt2_index) {
        ch_bowtie2_fasta    = Channel.value(
            [ [ id:'bowtie2_fasta' ], file(fasta_gz, checkIfExists: true) ]
        )
        ch_bowtie2_index    = Channel.value(
            [ [ id:'bowtie2_index' ], file(bt2_index, checkIfExists: true) ]
        )
    } else {
        ch_bowtie2_fasta    = Channel.value(
            [ [ id:'bowtie2_fasta' ], file(fasta_gz, checkIfExists: true) ]
        )
        ch_bowtie2_index   = null
    }

    //
    // MODULE: Create bowtie2 database from host FASTA
    //
    if (!ch_bowtie2_index) {
        BOWTIE2_BUILD(
            ch_bowtie2_fasta
        )
        ch_bowtie2_index    = BOWTIE2_BUILD.out.index
        ch_versions         = ch_versions.mix(BOWTIE2_BUILD.out.versions)
    }

    //
    // MODULE: Map, generate BAM with all reads and unmapped reads in FASTQ for downstream
    //
    BOWTIE2_ALIGN(
        fastq_gz,
        ch_bowtie2_index,
        ch_bowtie2_fasta,
        true,
        true
    )
    ch_versions = ch_versions.mix(BOWTIE2_ALIGN.out.versions)

    emit:
    fastq_gz    = BOWTIE2_ALIGN.out.fastq   // channel: [ val(meta), [ reads_1.fastq.gz, reads_2.fastq.gz ] ]
    bt2_log     = BOWTIE2_ALIGN.out.log     // channel: [ val(meta), bt2.log ]
    versions    = ch_versions               // channel: [ software_versions.yml ]
}

