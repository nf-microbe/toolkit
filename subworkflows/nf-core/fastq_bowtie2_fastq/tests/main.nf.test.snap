{
    "fasta input": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        [
                            "test1.unmapped_1.fastq.gz:md5,84b4ab15e6ffc9f29d0a3f1b79003744",
                            "test1.unmapped_2.fastq.gz:md5,40bbe14edd48a98cca6f53574b2198fd"
                        ]
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        [
                            "test2.unmapped_1.fastq.gz:md5,84b4ab15e6ffc9f29d0a3f1b79003744",
                            "test2.unmapped_2.fastq.gz:md5,40bbe14edd48a98cca6f53574b2198fd"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        "test1.bowtie2.log:md5,bd89ce1b28c93bf822bae391ffcedd19"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.bowtie2.log:md5,bd89ce1b28c93bf822bae391ffcedd19"
                    ]
                ],
                "2": [
                    "versions.yml:md5,3cd53b0efbbc7e2ce5673c1c97f046ee",
                    "versions.yml:md5,3cd53b0efbbc7e2ce5673c1c97f046ee",
                    "versions.yml:md5,bb1920373662263c1ba3d63171a55420"
                ],
                "bt2_log": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        "test1.bowtie2.log:md5,bd89ce1b28c93bf822bae391ffcedd19"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.bowtie2.log:md5,bd89ce1b28c93bf822bae391ffcedd19"
                    ]
                ],
                "fastq_gz": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        [
                            "test1.unmapped_1.fastq.gz:md5,84b4ab15e6ffc9f29d0a3f1b79003744",
                            "test1.unmapped_2.fastq.gz:md5,40bbe14edd48a98cca6f53574b2198fd"
                        ]
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        [
                            "test2.unmapped_1.fastq.gz:md5,84b4ab15e6ffc9f29d0a3f1b79003744",
                            "test2.unmapped_2.fastq.gz:md5,40bbe14edd48a98cca6f53574b2198fd"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3cd53b0efbbc7e2ce5673c1c97f046ee",
                    "versions.yml:md5,3cd53b0efbbc7e2ce5673c1c97f046ee",
                    "versions.yml:md5,bb1920373662263c1ba3d63171a55420"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-09T14:50:56.740439592"
    },
    "bt2 index input": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        [
                            "test1.unmapped_1.fastq.gz:md5,84b4ab15e6ffc9f29d0a3f1b79003744",
                            "test1.unmapped_2.fastq.gz:md5,40bbe14edd48a98cca6f53574b2198fd"
                        ]
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        [
                            "test2.unmapped_1.fastq.gz:md5,84b4ab15e6ffc9f29d0a3f1b79003744",
                            "test2.unmapped_2.fastq.gz:md5,40bbe14edd48a98cca6f53574b2198fd"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        "test1.bowtie2.log:md5,bd89ce1b28c93bf822bae391ffcedd19"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.bowtie2.log:md5,bd89ce1b28c93bf822bae391ffcedd19"
                    ]
                ],
                "2": [
                    "versions.yml:md5,3cd53b0efbbc7e2ce5673c1c97f046ee",
                    "versions.yml:md5,3cd53b0efbbc7e2ce5673c1c97f046ee"
                ],
                "bt2_log": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        "test1.bowtie2.log:md5,bd89ce1b28c93bf822bae391ffcedd19"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.bowtie2.log:md5,bd89ce1b28c93bf822bae391ffcedd19"
                    ]
                ],
                "fastq_gz": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        [
                            "test1.unmapped_1.fastq.gz:md5,84b4ab15e6ffc9f29d0a3f1b79003744",
                            "test1.unmapped_2.fastq.gz:md5,40bbe14edd48a98cca6f53574b2198fd"
                        ]
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        [
                            "test2.unmapped_1.fastq.gz:md5,84b4ab15e6ffc9f29d0a3f1b79003744",
                            "test2.unmapped_2.fastq.gz:md5,40bbe14edd48a98cca6f53574b2198fd"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3cd53b0efbbc7e2ce5673c1c97f046ee",
                    "versions.yml:md5,3cd53b0efbbc7e2ce5673c1c97f046ee"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-09T14:51:54.122661987"
    }
}